cibuildwheel/macos.py

```diff
@@ -34,21 +37,63 @@ def call(args: Sequence[PathOrStr], env: Optional[Dict[str, str]] = None, cwd: O
     return subprocess.check_call(args, env=env, cwd=cwd, shell=shell)
 
 
+def get_macos_version() -> Tuple[int, int]:
+    '''
+    Returns the macOS major/minor version, as a tuple, e.g. (10, 15) or (11, 0)
+
+    These tuples can be used in comparisons, e.g.
+        (10, 14) <= (11, 0) == True
+        (10, 14) <= (10, 16) == True
+        (11, 2) <= (11, 0) != True
+    '''
+    version_str, _, _ = platform.mac_ver()
+    version = tuple(map(int, version_str.split(".")[:2]))
+    return cast(Tuple[int, int], version)
+
+
 class PythonConfiguration(NamedTuple):
     version: str
     identifier: str
     url: str
 
 
-def get_python_configurations(
-        build_selector: BuildSelector) -> List[PythonConfiguration]:
+def get_python_configurations(build_selector: BuildSelector,
+                              architectures: Set[Architecture]) -> List[PythonConfiguration]:
 
     full_python_configs = read_python_configs('macos')
 
     python_configurations = [PythonConfiguration(**item) for item in full_python_configs]
 
-    # skip builds as required
-    return [c for c in python_configurations if build_selector(c.identifier)]
+    # filter out configs that don't match any of the selected architectures
+    python_configurations = [c for c in python_configurations
+                             if any(c.identifier.endswith(a.value) for a in architectures)]
+
+    # skip builds as required by BUILD/SKIP
+    python_configurations = [c for c in python_configurations if build_selector(c.identifier)]
+
+    # When running on macOS 11 and x86_64, the reported OS is '10.16', but
+    # there is no such OS - it really means macOS 11.
+    if get_macos_version() >= (10, 16):
+        if any(c.identifier.startswith('pp') for c in python_configurations):
+            # pypy doesn't work on macOS 11 yet
+            # See https://foss.heptapod.net/pypy/pypy/-/issues/3314
+            log.warning(unwrap('''
+                PyPy is currently unsupported when building on macOS 11. To build macOS PyPy wheels,
+                build on an older OS, such as macOS 10.15. To silence this warning, deselect PyPy by
+                adding "pp*-macosx*" to your CIBW_SKIP option.
+            '''))
+            python_configurations = [c for c in python_configurations if not c.identifier.startswith('pp')]
+
+        if any(c.identifier.startswith('cp35') for c in python_configurations):
+            # CPython 3.5 doesn't work on macOS 11
+            log.warning(unwrap('''
```

Comment:
This one doesn't bother me as a warning as much. I think the only place Python 3.5 is still around is in package managers for Linux.